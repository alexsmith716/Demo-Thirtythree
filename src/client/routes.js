import Root from '../containers/Root/Root';
import App from '../containers/App/App';
import Home from '../containers/Home/Home';
import NotFound from '../containers/NotFound/Loadable';

import AboutA from '../containers/AboutAAA/AboutA';
import AboutB from '../containers/AboutBBB/AboutB';
//  import About from '../containers/About/Loadable';
//  import GraphqlPage from '../containers/GraphqlPage/GraphqlPage';
import GraphqlPage from '../containers/GraphqlPage/Loadable';

//import { preloadData as preloadDataApp } from '../containers/App/preloadData';
//import { preloadData as preloadDataAboutThree } from '../containers/About/preloadData';

const routes = [
	{
		component: Root,
		path: '/',
		routes: [
			{
				path: '/',
				component: App,
				//  loadData: preloadDataApp,
				routes: [
					{
						path: '/',
						exact: true,
						component: Home,
					},
          {
          	path: '/aboutA',
          	exact: true,
          	component: AboutA,
          },
          {
            path: '/aboutB',
            exact: true,
            component: AboutB,
          },
					{
						path: '/graphqlpage',
						exact: true,
						component: GraphqlPage,
					},
					{
						path: '*',
						component: NotFound,
					},
				],
			},
		],
	},
];

export default routes;
