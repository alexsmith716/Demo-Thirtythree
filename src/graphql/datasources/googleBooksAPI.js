import { RESTDataSource } from 'apollo-datasource-rest';

const baseUrl = 'https://www.googleapis.com/books';

export class GoogleBooksAPI extends RESTDataSource {
	constructor() {
		super();
		this.baseURL = `${baseUrl}/v1/`
	}

	bookReducer(book) {
		return {
			id: book.id,
			cursor: `${book.id}`,
			title: book.volumeInfo.title,
			authors: book.volumeInfo.authors,
			publisher: book.volumeInfo.publisher,
			publishedDate: book.volumeInfo.publishedDate,
			description: book.volumeInfo.description,
			previewLink: book.volumeInfo.previewLink,
			smallThumbnail: book.volumeInfo.imageLinks && book.volumeInfo.imageLinks.smallThumbnail,
			favorite: false,
		};
	}

	// https://www.googleapis.com/books/v1/volumes?q=gmat&startIndex=0&orderBy=newest&projection=lite&maxResults=40
	// https://www.googleapis.com/books/v1/volumes/XDTlxgEACAAJ

	async getBooks(searchString, orderBy) {
		try {
			// const route = `volumes?q=${searchString.split(' ').join('+')}&startIndex=${startIndex}&orderBy=${orderBy}&projection=lite&maxResults=${maxResults}`;
			const route = `volumes?q=${searchString.split(' ').join('+')}&startIndex=0&orderBy=${orderBy}&projection=lite&maxResults=40`;
			// console.log('>>>>>>>>>>>>> googleBooksAPI > getBooks > route: ', route);
			const response = await this.get(route);
			// console.log('>>>>>>>>>>>>> googleBooksAPI > getBooks > response: ', response);
			console.log('>>>>>>>>>>>>> googleBooksAPI > getBooks > response: ', JSON.stringify(response));
			const reducedResponse = Array.isArray(response.items) ? response.items.map(book => this.bookReducer(book)) : [];
			console.log('>>>>>>>>>>>>> googleBooksAPI > getBooks > reducedResponse: ', reducedResponse);
			return reducedResponse;
		} catch (error) {
			console.error('>>>>>>>>>>>>> googleBooksAPI > getBooks > response > ERROR: ', error);
			return false;
		}
	}

	async getBook({id}) {
		try {
			const route = `volumes/${id}`;
			const response = await this.get(route);
			console.log('>>>>>>>>>>>>> googleBooksAPI > getBook > response: ', response);
			const reducedResponse = this.bookReducer(response);
			console.log('>>>>>>>>>>>>> googleBooksAPI > getBook > reducedResponse: ', reducedResponse);
			return reducedResponse;
		} catch (error) {
			console.error('>>>>>>>>>>>>> googleBooksAPI > getBook > response > ERROR: ', error);
			return false;
		}
	}
};
