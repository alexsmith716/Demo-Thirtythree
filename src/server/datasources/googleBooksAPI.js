const { RESTDataSource } = require('apollo-datasource-rest');

const baseUrl = 'https://www.googleapis.com/books';

class KaplanTestPrepBooks extends RESTDataSource {
	constructor() {
		super();
		this.baseURL = `${baseUrl}/v1/`
	}

	async getBooks(searchString) {
		const route = `volumes?q=${searchString.split(' ').join('+')}&projection=lite&maxResults=30`;
		console.log(`HTTP GET ${this.baseURL}${route}`);
		const data = await this.get(route);
		return data
	}
};

module.exports = { KaplanTestPrepBooks };
