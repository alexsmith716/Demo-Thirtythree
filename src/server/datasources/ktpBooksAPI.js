const { RESTDataSource } = require('apollo-datasource-rest');

class KtpBooksAPI extends RESTDataSource {
	constructor() {
		super();
		this.baseURL = 'https://www.googleapis.com/books/v1/volumes?q=kaplan%20test%20prep';
	}

};

module.exports = KtpBooksAPI;
