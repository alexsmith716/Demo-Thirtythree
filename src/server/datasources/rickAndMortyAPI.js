const { RESTDataSource } = require('apollo-datasource-rest');

class RickAndMortyAPI extends RESTDataSource {
	constructor() {
		super();
		this.baseURL = 'https://rickandmortyapi.com/api/character';
	}

	async character({ id }) {
		const data = await this.get('/' + id)
		return data
	}
};

module.exports = RickAndMortyAPI;
