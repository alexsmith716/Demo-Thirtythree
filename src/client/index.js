import 'core-js/stable';
import 'regenerator-runtime/runtime';
import React from 'react';
import ReactDOM from 'react-dom';
import { Router } from 'react-router';
import { renderRoutes } from 'react-router-config';
import { createBrowserHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { loadableReady } from '@loadable/component';
import localForage from 'localforage';
import { getStoredState } from 'redux-persist';
import { Provider } from 'react-redux';
import { ApolloProvider, ApolloClient, createHttpLink, InMemoryCache, ApolloLink } from '@apollo/client';
import { onError } from '@apollo/client/link/error';

import asyncGetPromises from '../utils/asyncGetPromises';
import { RouterTrigger } from '../components/RouterTrigger/RouterTrigger';
import routes from './routes';
import { apiClient } from '../helpers/apiClient';
import configureStore from '../redux/configureStore';
import isOnline from '../utils/isOnline';
import '../styled/fonts.css';

import { ThemeContext } from '../styled/ThemeContext';

const persistConfig = {
	key: 'root',
	storage: localForage,
	// redux-persist:
	// inboundState:  the state being rehydrated from storage
	// originalState: the state before the REHYDRATE action
	stateReconciler(inboundState, originalState) {
		// preloadedState from window object
		return originalState;
	},
	// redux-persist:
	// whitelist: ['info', 'infoAlert', 'infoAlertThree',],
};

const spinnerContainer = document.createElement('div');
spinnerContainer.classList.add('spinner-progress');
const dest = document.getElementById('react-root');
document.body.insertBefore(spinnerContainer, dest);

const client = apiClient();

const providers = {
	client,
};

// =====================================================

const render = async () => {
	// redux-persist:
	// delays rendering of app UI until persisted state has been retrieved and saved to redux
	const preloadedState = await getStoredState(persistConfig);
	const online = window.REDUX_DATA ? true : await isOnline();
	const history = createBrowserHistory();

	const store = configureStore({
		history,
		data: {
			...preloadedState,
			...window.REDUX_DATA,
			online,
		},
		helpers: providers,
		persistConfig,
	});

	// =====================================================

	const httpLink = createHttpLink({
		uri: 'https://rickandmortyapi.com/graphql/',
	});

	const errorLink = onError(({ graphQLErrors, networkError }) => {
		if (graphQLErrors) {
			graphQLErrors.map(({ message, locations, path }) =>
				console.log(`>>>> CLIENT > [GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`,),
			);
		}

		if (networkError) {
			console.log(`>>>> CLIENT > [Network error]: ${networkError}`);
		}
	});

	const link = ApolloLink.from([
		errorLink,
		httpLink,
	]);

	// https://github.com/apollographql/apollo-client/blob/main/docs/source/pagination/core-api.mdx
	// https://github.com/apollographql/apollo-client/docs/source/local-state/reactive-variables.mdx
	// https://github.com/apollographql/apollo-client/docs/source/local-state/managing-state-with-field-policies
	const clientCache = new InMemoryCache({
		typePolicies: {
			// =====================================================
			Query: {
				fields: {
					// --------------------------------------
					// define field policy for "googleBooksList"
					// define field policy to merge the results of paginated queries into a single list
					// googleBooksList: relayStylePagination(),
					googleBooksList: {
						keyArgs: false,
						merge(existing, incoming) {
							//console.log('>>>> CLIENT > InMemoryCache > existing AAAAAAAA: ', existing);
							//console.log('>>>> CLIENT > InMemoryCache > incoming BBBBBBBB: ',  incoming);
							let books = [];
							if (existing && existing.books) {
								books = books.concat(existing.books);
							}
							if (incoming && incoming.books) {
								books = books.concat(incoming.books);
							}
							const q = {...incoming}
							const w = {books}
							const x = {...incoming, books}
							//console.log('>>>> CLIENT > InMemoryCache > incoming CCCCCCCCCCC: ',  q);
							//console.log('>>>> CLIENT > InMemoryCache > books DDDDDDDDDDD: ',  w);
							//console.log('>>>> CLIENT > InMemoryCache > incoming EEEEEEEE: ', x);
							return {
								...incoming,
								books,
							};
						}
					}
				}
			}
		}
	});

	const clientApollo = new ApolloClient({
		ssrMode: false,
		cache: clientCache.restore(window.__APOLLO_STATE__),
		link,
	});

	// =====================================================

	const triggerHooks = async (hydrateRoutes, pathname) => {
		spinnerContainer.classList.add('spinner');
		if (window.__PRELOADED__) {
			delete window.__PRELOADED__;
		} else {
			await asyncGetPromises(hydrateRoutes, pathname, store);
		}
		spinnerContainer.classList.remove('spinner');
	};

	const hydrate = (hydrateRoutes) => {
		const element = (
			<HelmetProvider>
				<ApolloProvider client={clientApollo}>
					<Provider store={store}>
						<Router history={history}>
							<ThemeContext>
								<RouterTrigger triggerProp={(pathname) => triggerHooks(hydrateRoutes, pathname)}>
									{renderRoutes(hydrateRoutes)}
								</RouterTrigger>
							</ThemeContext>
						</Router>
					</Provider>
				</ApolloProvider>
			</HelmetProvider>
		);


		if (dest && dest.innerHTML !== "") {
			console.log('############## CLIENT > ReactDOM.hydrate 00000 ###########')
		} else {
			console.log('############## CLIENT > ReactDOM.render 00000 ###########')
		}

		if (dest.hasChildNodes()) {
			console.log('############## CLIENT > ReactDOM.hydrate 111111 ###########')
			ReactDOM.hydrate(element, dest);
		} else {
			console.log('############## CLIENT > ReactDOM.render 111111 ###########')
			ReactDOM.render(element, dest);
		}
	};

	hydrate(routes);
};

loadableReady(() => {
	render();
});
