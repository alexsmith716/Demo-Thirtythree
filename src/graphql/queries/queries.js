import { gql } from '@apollo/client';
import { fragmentTypeDroid } from '../fragments/fragments';

export const GET_HELLO = gql`
	{
		hello
	}
`;

export const GET_KTP_BOOKS_REST = gql`
	query GetKptBooksRest($search: String!) {
		search(searchString: $search) {
			books {
				id
				title
				authors
				publisher
				publishedDate
        description
				imageLinks {
					smallThumbnail
				}
			}
		}
	}
`;

export const GET_CHARACTER_REST = gql`
	query GetCharacterRest($id: ID) {
		character(id: $id) {
			id
			name
			status
			species
			type
			gender
		}
	}
`;

export const GET_REVIEWS = gql`
	query GetReviews($episode: Episode!) {
		reviews(episode: $episode) {
			episode
			stars
			commentary
		}
	}
`;

export const GET_A_DROID = gql`
	query GetADroid($droidID: ID!) {
		droid(id: $droidID) {
			id
			name
			friends {
				id
				name
			}
			appearsIn
			primaryFunction
		}
	}
`;

export const GET_A_DROID_ALIAS = gql`
	query GetADroid($droidIDa: ID!, $droidIDb: ID!) {
		droidIDa: droid(id: $droidIDa) {
			...fragmentTypeDroid
		}
		droidIDb: droid(id: $droidIDb) {
			...fragmentTypeDroid
		}
	}
	${fragmentTypeDroid}
`;

export const GET_CHARACTER = gql`
	query Character($id: ID){
		character(id: "1") @rest(type: "Post", path: "character/1/") {
			id
			name
			status
			species
			type
			gender
			origin {
				name
				type
				dimension
			}
			location {
				name
				type
				dimension
			}
			image
			episode {
				name
				episode
			}
		}
	}
`;

export const ADD_REVIEW = gql`
	mutation createReview($episode: Episode, $review: ReviewInput!) {
		createReview(episode: $episode, review: $review ) {
			episode
			stars
			commentary
		}
	}
`;

export const JUST_GET_REVIEWS = gql`
	query GetEpisodeReviews($episode: Episode!) {
		reviews(episode: $episode)
	}
`;

export const GET_HERO = gql`
	{
		hero {
			name
		}
	}
`;

export const GET_THE_SCHEMA = gql`
	{
		__schema {
			types {
				name
				kind
				description
				fields {
					name
				}
			}
		}
	}
`;
