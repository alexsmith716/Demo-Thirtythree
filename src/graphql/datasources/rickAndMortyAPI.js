import { DataSource } from 'apollo-datasource';
import fetch from 'isomorphic-fetch';

// import graphqlClient from '../../apollo/graphqlClient';
import { GET_RICK_AND_MORTY_CHARACTER, GET_RICK_AND_MORTY_CHARACTERS, GET_RICK_AND_MORTY_CHARACTERS_BY_IDS } from '../queries/queries.js';

export class RickAndMortyAPI extends DataSource {
	constructor() {
		super();
		this.baseURL = 'https://rickandmortyapi.com/graphql';
	}

	graphqlClient({ query, variables={}, method='POST' }) {
    console.log('#################### RickAndMortyAPI/graphqlClient/characters ####################')
		return fetch('https://rickandmortyapi.com/graphql', {
			method: method,
			headers: {
				'Content-Type': 'application/json',
			},
			body: JSON.stringify({
				query: query,
				variables: {...variables}
			}),
		})
			.then(response => response.json())
			.then(data => {
				if (data.error) {
					throw new Error(data.error);
				}
				// console.log('>>>> RickAndMortyAPI > DATA:', JSON.stringify(data));
				return data;
			})
			.catch(err => {
				console.error('>>>> RickAndMortyAPI > ERROR:', err.message);
				return null;
			});
	};

	async getCharacter({id}) {
		try {
			const response = await this.graphqlClient({query: GET_RICK_AND_MORTY_CHARACTER, variables: {id: id}});
			const { data: { character }} = response;
			return character;
		} catch (error) {
			console.error('>>>>>>>>>>>>> RickAndMortyAPI > Query > getCharacter > ERROR: ', error);
			return false;
		}
	}

	async getCharacters({page, filter}) {
    console.log('#################### RickAndMortyAPI/characters ####################')
		try {
			const response = await this.graphqlClient({query: GET_RICK_AND_MORTY_CHARACTERS, variables: {page: page, filter: filter}});
			const { data: { characters }} = response;
			return characters;
		} catch (error) {
			console.error('>>>>>>>>>>>>> RickAndMortyAPI > Query > getCharacters > ERROR: ', error);
			return false;
		}
	}

	async getCharactersByIds({ids}) {
		try {
			const response = await this.graphqlClient({query: GET_RICK_AND_MORTY_CHARACTERS_BY_IDS, variables: {ids: ids}});
			const { data: { charactersByIds }} = response;
			return charactersByIds;
		} catch (error) {
			console.error('>>>>>>>>>>>>> RickAndMortyAPI > Query > getCharacters > ERROR: ', error);
			return false;
		}
	}
};
