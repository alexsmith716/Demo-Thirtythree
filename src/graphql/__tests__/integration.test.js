import { createTestClient } from 'apollo-server-testing';
import nock from'nock';

// integration test being run against a test instance of apollo-server

// =========================
import { toPromise } from '@apollo/client';
// =========================

import { GET_GOOGLE_BOOK } from '../queries/queries.js';
import { GoogleBooksAPI } from '../datasources/googleBooksAPI';

import { constructTestServer, startTestServer } from './__utils';

// the mocked REST API data
import { mockBookResponse } from '../datasources/__tests__/googleBooksAPI.test.js';

describe('Queries', () => {
	it('fetches single book', async () => {
		const {server, googleBooks} = constructTestServer();

		googleBooks.get = jest.fn(() => mockBookResponse);

		const {query} = createTestClient(server);

		const res = await query({query: GET_GOOGLE_BOOK, variables: {id: 'uW_zzQEACAAJ'}});

		expect(res).toMatchSnapshot();
	});
});

//	describe('Server', () => {
//		let stop, graphql;
//	
//		beforeEach(async () => {
//			const testServer = await startTestServer();
//			stop = testServer.stop;
//			graphql = testServer.graphql;
//		});
//	
//		afterEach(() => stop());
//	
//		it('gets a single book', async () => {
//			const res = await toPromise(graphql({ query: GET_GOOGLE_BOOK, variables: { id: 'uW_zzQEACAAJ' } }));
//	
//			expect(res).toMatchSnapshot();
//		});
//	});
