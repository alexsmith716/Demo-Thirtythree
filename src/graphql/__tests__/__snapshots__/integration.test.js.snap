// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetches single book 1`] = `
Object {
  "data": Object {
    "googleBook": Object {
      "authors": Array [
        "Kaplan Test Prep",
      ],
      "description": "<b>Kaplan’s<i> MCAT Physics and Math Review 2022–2023 </i>offers an expert study plan, detailed subject review, and hundreds of online and in-book practice questions—all authored by the experts behind the MCAT prep course that has helped more people get into medical school than all other major courses combined.</b><br><br> Prepping for the MCAT is a true challenge. Kaplan can be your partner along the way—offering guidance on where to focus your efforts and how to organize your review. This book has been updated to match the AAMC’s guidelines precisely—no more worrying about whether your MCAT review is comprehensive!<br><br><i>The Most Practice</i><br><ul><li>More than 350 questions in the book and access to even more online—more practice than any other MCAT physics and math book on the market.</li></ul><br><i>The Best Practice</i><br><ul><li>Comprehensive physics and math subject review is written by top-rated, award-winning Kaplan instructors.</li><li>Full-color, 3-D illustrations from Scientific American, charts, graphs and diagrams help turn even the most complex science into easy-to-visualize concepts.</li><li>All material is vetted by editors with advanced science degrees and by a medical doctor.</li><li>Online resources, including a full-length practice test, help you practice in the same computer-based format you’ll see on Test Day.</li></ul><br><i>Expert Guidance</i><br><ul><li>High-yield badges throughout the book identify the top 100 topics most tested by the AAMC.</li><li>We know the test: The Kaplan MCAT team has spent years studying every MCAT-related document available. </li><li>Kaplan’s expert psychometricians ensure our practice questions and study materials are true to the test.</li></ul>",
      "favorite": false,
      "id": "uW_zzQEACAAJ",
      "publishedDate": "2021-07-06",
      "publisher": "Kaplan Publishing",
      "smallThumbnail": "http://books.google.com/books/content?id=uW_zzQEACAAJ&printsec=frontcover&img=1&zoom=5&imgtk=AFLRE72XJEFAGe-1GZaWa1lIQk-Dfb11hClDf5V94UbaQamcVQLXrkyl_DTjFvc2t9Nc4cPPTSpUXaRzGOJpIoKlSqKLB4Nd1JygtR_mh_-6ehnH5Y2r0JB0VZI5So0fErq9vnNa1wHD&source=gbs_api",
      "title": "MCAT Physics and Math Review 2022-2023",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
